website/addons/base/__init__.py

```diff
@@ -166,17 +171,123 @@ class GuidFile(GuidStoredObject):
     }
 
     @property
-    def file_url(self):
+    def provider(self):
+        raise NotImplementedError
+
+    @property
+    def version_identifier(self):
         raise NotImplementedError
 
     @property
+    def unique_identifier(self):
+        raise NotImplementedError
+
+    @property
+    def guid_url(self):
+        return '/{0}/'.format(self._id)
+
+    @property
+    def name(self):
+        return self._metadata_cache['name']
+
+    @property
+    def file_name(self):
+        if self.revision:
+            return '{0}_{1}.html'.format(self._id, self.revision)
+        return '{0}_{1}.html'.format(self._id, self.unique_identifier)
+
+    @property
+    def joinable_path(self):
+        if self.path.startswith('/'):
+            return self.path[1:]
+        return self.path
+
+    @property
+    def _base_butler_url(self):
+        url = furl.furl(settings.WATERBUTLER_URL)
+
+        url.args.update({
+            'nid': self.node._id,
+            'path': self.path,
+            'provider': self.provider,
+            'cookie': request.cookies.get(settings.COOKIE_NAME)
+        })
+
+        if self.revision:
+            url.args[self.version_identifier] = self.revision
+
+        return url
+
+    @property
+    def download_url(self):
+        url = self._base_butler_url
+        url.path.add('file')
+        return url.url
+
+    @property
+    def metadata_url(self):
+        url = self._base_butler_url
+        url.path.add('data')
+
+        return url.url
+
+    @property
+    def cache_path(self):
```

Comment:
To be more explicit, how about prefixing methods that are mfr-specific with mfr? e.g. `mfr_cache_path`